File: lab5_dist_normal.R
# Title     : Distribuicao Normal
# Objective : Apresentar as ferramentas em R para suporte a aplicacoes com Distribuicao Normal
# Created by: accol
# Created on: 23/05/2020

# Calculo da distribuicao normal no R
# P[X < x]
# prnorm()
# P[X > x] => lower.tail = FALSE
# qnorm() => faz o processo inverso
# rnorm[] => gera numeros aleatorios normalmente distribuidos
# qqnorm[] => gera o teste de normalidade
# qqline[] => gera a linha de melhor ajuste (completando o qnorm) para comparacao
# shapiro.test[] => realiza o teste de SHAPIRO-WILK

# Neste exemplo temos uma caixa com objetos. Sabe-se que a media do peso e de  8 quilos, o desvio padrao e 2,
# e estao distribuidos de forma Normal => pede-se a chance de um objeto retirado da caixa pesar menos
# de 6 quilos ||> para esse exemplo simples, devemos lembrar que a probabilidade e cumulativa e que estamos
# trabalhando na regiao a esquerda da media (neste caso, nao precisamos subtrair de um)

# Vamos usar a funcao pnorm() que exige tres paraetros => o primeiro o que estou buscando >- no caso 6 quilos
# abaixo, o segundo parametro e a media e o terceiro o desvio padrao ||> problema do tipo menor que

menor_seis = pnorm(6, 8, 2)
print('A probabilidade de retirar um objeto com menos de 6 quilos e')
print(menor_seis)

# Neste segundo problema queremos saber qual a chance de retiramos um objeto com mais de 6 quilos ||> fique atento
# e acompanhe => trata-se de um problema do tipo maior que => lembra do parametro Lower Tail que normalmente vem
# setado como True

maior_seis = pnorm(6, 8, 2, lower.tail = F)
print('A probabilidade de retirar um objeto com mais de 6 quilos e')
print(maior_seis)

# Neste terceiro problema queremos saber qual a chance de retiramos um objeto com menos de 6 quilos ou mais
# de 10 quilos||> fique atento a clausula 'ou' e acompanhe => temos que encontrar a probabilidade de ser menor
# de 6 quilos e depois maior de 10 quilos, por fim somar as duas, ou entao, neste caso, basta encontrar uma e
# multiplicar por 2

menor_seis_maior_dez = pnorm(6, 8, 2) + pnorm(10, 8, 2, lower.tail = F)
print('A probabilidade de retirar um objeto com menos de 6 quilos ou mais de 10 quilos e:')
print(menor_seis_maior_dez)

# Nosso quarto problema nos pede para encontrarmos a probabilidade de retirarmos um objeto com menos de 10 quilos
# e com mais de 8 quilos ||> Note que estamos trabalhando com duas probabilidades na cauda a esquerda => neste
# caso, encontramos cada uma e subtraimos uma da outra, observe

menor_10_maior_8 = pnorm(10, 8, 2) - pnorm(8, 8, 2)
print('A probabilidade de retirar um objeto com menos de 10 quilos e mais de 8 quilos e:')
print(menor_10_maior_8)

# <$> AGORA VAMOS AOS TESTES >=> Queremos saber se uma distribuicao e ou nao normal <$>
# Primeiramente atraves do Histograma
# Para este exemplo, primeiramente vamos gerar dados uniformente distribuidos utilizando a funcao >- rnorm()

x = rnorm(100) # devera retornar 100 numeros seguindo uma distribuicao normal
print('Variavel gerada atraves de rnorm():')
print(x)

# Gerando o grafico de normalidade => primeiro usando qqnorm()
print('Gerando o grafico de normalidade, para inspecao visual')
print(qqnorm(x))
# Gerando o grafico de normalidade => primeiro usando qqline()
print('Gerando o grafico de linha, para melhorar a inspecao visual')
print(qqline(x))
# Gerando um histograma => primeiro usando hist()
print('Gerando historgrama, para melhorar a inspecao visual')
print(hist(x, main = 'Histograma da Distribuicao de X', xlab = 'Dados', ylab = 'Frequencia'))
# Usando agora o teste de Shapiro-Wilk |||> shapiro.test()
print('Usando o teste de Shapiro-Wilk => com os parametros normais')
print(shapiro.test(x))
# Note que pelos resultados estamos claramente com uma distribuicao NORMAL -> basta comparar o valor de p
# com seu valor de Alfa

File: lab5_dist_normal.py
# Title     : Distribuicao Normal
# Objective : Apresentar as ferramentas em Python para suporte a aplicacoes com Distribuicao Normal
# Created by: accol
# Created on: 23/05/2020

# Calculo da distribuicao normal no Python


from scipy import stats
from scipy.stats import norm
import matplotlib.pyplot as plt


# Conjunto de objetos em uma cesta, a media e 8 e o desvio padrao e 2 => Queremos saber qual a probabilidade
# de tiramos um objeto ao acaso e seu peso ser inferior a 6 quilos?
# Vamos usar a função norm.cdf() -> usada quando buscamos por um valor do tipo "menor que" tem como primeiro parâmetro o valor alvo (no caso 6 quilos), o segundo parâmetro é a média e o terceiro o desvio padrão

menor_seis = norm.cdf(6, 8, 2)
print('\nA probabilidade de retirar um objeto com menos de 6 quilos da cesta é: ', menor_seis)

# Neste problema, queremos saber a probabilidade de retiramos um objeto que o peso seja maior que 6 quilos? Quando usamos o conceito de maior, precisamos fazer uso de outra funcao >- norm.sf() => usamos os mesmos três artumentos

maior_seis = norm.sf(6, 8, 2)
print('\nA probabilidade de retirar um objeto com mais de 6 quilos da cesta é: ', maior_seis)

# Neste problema, queremos saber a probabilidade de retiramos um objeto que o peso é menor que 6 quilos ou maior que 10 quilos? Quando usamos o conceito de menor ou maior, precisamos fazer um somatório das probabilidades

menor_seis_maior_dez = norm.cdf(6, 8, 2) + norm.sf(10, 8, 2)
print('\nA probabilidade de retirar um objeto com mais de 6 quilos da cesta é: ', menor_seis_maior_dez)

# Neste problema, queremos saber a probabilidade de retiramos um objeto que o peso seja menor que 10 quilos e maior que 8 quilos? Note a falta da cláusula "ou", neste caso é preciso estar "entre" 8 "e" 10 ||> precisamos subtrair. Fique atento, pois embora, queremos que esteja entre 8 e 10, vamos subtrair menor de 10, por menor de 8. Para melhor entender pense no gráfico de sino

menor_dez_maior_oito = norm.cdf(10, 8, 2) - norm.cdf(8, 8,2)
print('\nA probabilidade de retirar um objeto com menos de 10 e mais de 8 quilos da cesta é: ', menor_dez_maior_oito)
# Caso prefira você poderá calcular considerando o uso do norm.sf() para os valores maior que
menor_dez_maior_oito_2 = norm.cdf(10, 8, 2) - norm.sf(8, 8,2)
print('\nA probabilidade de retirar um objeto com menos de 10 e mais de 8 quilos da cesta é: ', menor_dez_maior_oito_2)

# <$> AGORA FAREMOS TESTES PARA VERIFICAR SE A DISTRIBUICAO É OU NÃO NORMAL <$>
# Primeiro vamos criar uma variável para recener uma distribuição Normal gerada através de norm.rvs(size)
dados = norm.rvs(size=100)
print('\nOs dados gerados a partir do gerador de distribuição normal é:\n')
print(dados)
print(type(dados))
# Gerando o gráfico de distribuição Normal
g_normal = stats.probplot(dados, plot=plt)
plt.show(g_normal)
# Uma inspeção visual simples nos permite verificar que se trata de uma distribuição Normal

# Para confirmar faremos o teste de Shapiro-Wilk
shapiro_teste_hipotese = stats.shapiro(dados)
print('\nO resultado do teste de hipotese de Shapiro é: \n')
print(shapiro_teste_hipotese)
# Os resultados do teste de Shapiro confirmam a distribuição Normal dos dados
